Token22 batch 3 (#375)
Unboxed-Software/solana-course: content/transfer-hook.md

cli
3/12
Saved

Hide assistant

> This suggestion is part of Grammarly Premium.

Reviewer: Fixed in latest pushes

---

title: Transfer Hook

objectives:

- Create a program that applies the "transfer-hook" interface

- Create a mint with a transfer hook

- Transfer a token with a transfer hook successfully

---

# Summary

- The `transfer hook` extension allows developers to run custom logic on their tokens on every transfer

- When a token has a transfer hook, the Token Extensions Program will invoke the transfer hook instruction on every token transfer

- For the program to be able to act as a transfer hook program, it needs to implement the `TransferHook` interface

> - Some transfer hooks will require additional accounts, if this is the case, every transfer instruction will have to pass them. These are defined in `extra-account-metas` on the transfer hook extension.

Reviewer: Fixed in latest pushes

- Within the transfer hook CPI, the sender, mint, receiver and owner are all de-escalated, meaning they are read-only to the hook. Meaning none of those accounts can sign or be written to.

# Overview

> The `transfer-hook` extension allows custom on-chain logic to be after each transfer within the same transaction. More specifically, the `transfer-hook` extension requires a 'hook' or 'callback' in the form of a Solana program following the [Transfer Hook Interface](https://github.com/solana-labs/solana-program-library/tree/master/token/transfer-hook/interface). Then every time any token of that mint is transferred the Token Extensions Program calls this 'hook' as a CPI.

Reviewer: Fixed in latest pushes

Additionally, the `transfer-hook` extension also stores `extra-account-metas`, which are any additional accounts needed for the hook to function. 

> This extension opens up the door to unique use cases. One use case it's great for is acting as a token guard, only allowing tokens to be transferred if specific requirements are met.

Reviewer: Fixed in latest pushes

> In this lesson, we'll explore how to implement transfer hooks on-chain and work with them off-chain.

Reviewer: Fixed in latest pushes

> In this lesson, we'll explore how to implement transfer hooks on-chain and work with them off-chain.

Reviewer: Fixed in latest pushes

## Implementing transfer hooks onchain

The first part of creating a mint with a `transfer hook` is to find or create an on-chain program that follows the [Transfer Hook Interface](https://github.com/solana-labs/solana-program-library/tree/master/token/transfer-hook/interface).

> The Transfer Hook Interface specifies the following [instructions](https://github.com/solana-labs/solana-program-library/blob/master/token/transfer-hook/interface/src/instruction.rs):

Reviewer: Fixed in latest pushes

- `Execute`: An instruction that the Token Extensions Program invokes on every token transfer

- `InitializeExtraAccountMetaList` (optional): Creates an account (`extra_account_meta_list`) that stores a list of additional accounts required by the `Execute` instruction

> - `UpdateExtraAccountMetaList` (optional): Updates the list of additional accounts by overwriting the existing list

Reviewer: Fixed in latest pushes

Technically it's not required to implement the `InitializeExtraAccountMetaList` instruction using the interface, but it's still required to have the `extra_account_meta_list` account. This account can be created by any instruction on a Transfer Hook program. However, the Program Derived Address (PDA) for the account must be derived using the following seeds:

- The hard-coded string `extra-account-metas`

- The Mint Account address

- The Transfer Hook program ID

> ```js

Reviewer: Fixed in latest pushes

const [pda] = PublicKey.findProgramAddressSync(

  [

    Buffer.from("extra-account-metas"),

    mint.publicKey.toBuffer()

  ],

  program.programId, // transfer hook program ID

);

```

By storing the extra accounts required by the `Execute` instruction in the `extra_account_meta_list` PDA, these accounts can be automatically added to a token transfer instruction from the client. We'll see how to do that in the off-chain section.

### 1. `initialize_extra_account_meta_list` instruction:

> When we transfer a token using the Token Extensions Program, the program will examine our mint to determine if it has a transfer hook. If a transfer hook is present, the Token Extensions Program will initiate a CPI (cross-program invocation) to our transfer hook program. The Token Extensions Program will then pass all the accounts in the transfer (including the extra accounts specified in the `extra_account_meta_list`) to the transfer hook program. However, before passing the 4 essential accounts (sender, mint, receiver, owner), it will de-escalate them (i.e. Remove the mutable or signing abilities for security reasons).

Reviewer: Fixed in latest pushes

The `extra_account_meta_list` has to be created before any transfer occurs. It's also worth noting that we can update the list of accounts in the `extra_account_meta_list` by implementing and using the `UpdateExtraAccountMetaList` instruction if necessary.

The `extra_account_meta_list` is just a list of `ExtraAccountMeta`. Let's take a look at the struct `ExtraAccountMeta` [in the source code](https://github.com/solana-labs/solana-program-library/blob/4f1668510adef2117ee3c043bd26b79789e67c8d/libraries/tlv-account-resolution/src/account.rs#L90):

```rust

impl ExtraAccountMeta {

    /// Create a `ExtraAccountMeta` from a public key

    /// This represents standard `AccountMeta`

    pub fn new_with_pubkey(

        pubkey: &Pubkey,

        is_signer: bool,

        is_writable: bool,

    ) -> Result<Self, ProgramError> {

        Ok(Self {

            discriminator: 0,

            address_config: pubkey.to_bytes(),

            is_signer: is_signer.into(),

            is_writable: is_writable.into(),

        })

    }

    /// Create an `ExtraAccountMeta` from a list of seed configurations

>     /// This represents a PDA

Reviewer: Fixed in latest pushes

    pub fn new_with_seeds(

        seeds: &[Seed],

        is_signer: bool,

        is_writable: bool,

    ) -> Result<Self, ProgramError> {

        Ok(Self {

            discriminator: 1,

            address_config: Seed::pack_into_address_config(seeds)?,

            is_signer: is_signer.into(),

            is_writable: is_writable.into(),

        })

    }

    /// Create an `ExtraAccountMeta` from a list of seed configurations, representing

>     /// a PDA for an external program

Reviewer: Fixed in latest pushes

    ///

    /// This PDA belongs to a program elsewhere in the account list, rather

    /// than the executing program. For a PDA on the executing program, use

    /// `ExtraAccountMeta::new_with_seeds`.

    pub fn new_external_pda_with_seeds(

        program_index: u8,

        seeds: &[Seed],

        is_signer: bool,

        is_writable: bool,

    ) -> Result<Self, ProgramError> {

        Ok(Self {

            discriminator: program_index

                .checked_add(U8_TOP_BIT)

                .ok_or(AccountResolutionError::InvalidSeedConfig)?,

            address_config: Seed::pack_into_address_config(seeds)?,

            is_signer: is_signer.into(),

            is_writable: is_writable.into(),

        })

    }

```

We have three methods for creating an `ExtraAccountMeta`:

> 1. `ExtraAccountMeta::new_with_pubkey` - For any normal account ( Not a program account )

Reviewer: Fixed in latest pushes

2. `ExtraAccountMeta::new_with_seeds` - For a program account PDA from the calling transfer hook program 

3. `ExtraAccountMeta::new_external_pda_with_seeds` - For a program account PDA from a different external program

Now that we know the accounts we can store them in `extra_account_meta_list`. Let's talk about the `InitializeExtraAccountMetaList` instruction itself. For most implementations, it should simply just create the `extra_account_meta_list` account and load it up with any additional accounts it needs. 

Let's take a look at a simple example where we'll initialize an `extra_account_meta_list` with two additional arbitrary accounts, `some_account` and a `pda_account`. The `initialize_extra_account_meta_list` function will do the following:

> 1. Prepare the accounts that we need to store in the `extra_account_meta_list` account as a vector (we'll talk about that in-depth in a sec).

Reviewer: Fixed in latest pushes

2. Calculate the size and rent required to store the list of `ExtraAccountMetas`.

3. Make a CPI to the System Program to create an account and set the Transfer Hook Program as the owner, and then initialize the account data to store the list of `ExtraAccountMetas`.

```rust

#[derive(Accounts)]

pub struct InitializeExtraAccountMetaList<'info> {

  #[account(mut)]

>   payer: Signer<'info>,

Reviewer: Fixed in latest pushes

  /// CHECK: ExtraAccountMetaList Account, must use these seeds

  #[account(

        mut,

        seeds = [b"extra-account-metas", mint.key().as_ref()],

        bump

    )]

>   pub extra_account_meta_list: AccountInfo<'info>,

Reviewer: @jamesrp13

>   pub extra_account_meta_list: AccountInfo<'info>,

Reviewer 2: This doesn't appear to be specific to the rust snippets. Something must have gone wrong copying the file over at some point because there are extra newlines everywhere. Pushing a commit now that removes thousands of empty lines 😂

    ///     * 1 - Start index of account data

    ///     * 1 - Length of account data starting at index

    AccountData {

        /// The index of the account in the entire accounts list

        account_index: u8,

        /// The index where the bytes of an account data argument begin

        data_index: u8,

        /// The length of the argument (number of bytes)

        ///

        /// Note: Max seed length is 32 bytes, so `u8` is appropriate here

        length: u8,

    },

}

```

As we can see from the code above, there are four main ways to provide seeds:

1. A literal hard-coded argument, such as the string `"some-seed"`.

2. An instruction-provided argument, to be resolved from the instruction data. This can be done by giving the start index and the length of the data we want to have as a seed.

3. The public key of an account from the entire accounts list. This can be done by giving the index of the account (we'll talk about this more soon).

4. An argument to be resolved from the inner data of some account. This can be done by giving the index of the account, the start index of the data, along with the length of the data we want to have as a seed.

> To use the 2 last methods of setting the seed, you need to get the account index. This represents the index of the account passed into the `Execute` function of the hook. This is standardized. 

Reviewer: Fixed in latest pushes

- index 0-3 will always be, `source`, `mint`, `destination`, and `owner` respectively

- index 4: will be the `extra_account_meta_list`

- index 5+: will be in whatever order you create your `account_metas`

```rust

    // index 0-3 are the accounts required for token transfer (source, mint, destination, owner)

    // index 4 is the extra_account_meta_list account

  let account_metas = vec![

    // index 5 - some_account

    ExtraAccountMeta::new_with_pubkey(&ctx.accounts.some_account.key(), false, true)?, 

    // index 6 - pda_account

    ExtraAccountMeta::new_with_seeds(

      &[

        Seed::Literal {

          bytes: "some-seed".as_bytes().to_vec(),

        },

      ],

      true, // is_signer

      true // is_writable

    )?,

  ];

```

> Now, let's say that the `pda_account` was PDA'd off of "some-seed" and `some_account`. This is where we can specify the account key index:

Reviewer: Fixed in latest pushes

```rust

  // index 0-3 are the accounts required for token transfer (source, mint, destination, owner)

  // index 4 is the extra_account_meta_list account

  let account_metas = vec![

    // index 5 - some_account

    ExtraAccountMeta::new_with_pubkey(&ctx.accounts.some_account.key(), false, true)?, 

    // index 6 - pda_account

    ExtraAccountMeta::new_with_seeds(

      &[

        Seed::AccountKey {

          index: 5, // index of `some_account`

        },

        Seed::Literal {

          bytes: "some-seed".as_bytes().to_vec(),

        },

      ],

      true, // is_signer

      true // is_writable

    )?,

  ];

```

> Note: Remember that the accounts indexed 0-4 are defined by the `Execute` function of the transfer hook. They are: `source`, `mint`, `destination`, `owner`, `extra_account_meta_list` respectively. The first four of which, are de-escalated, or read-only. These will always be read-only. If you try to be sneaky and add any of these first four accounts into the `extra_account_meta_list`, they will always be interpreted as read-only, even if you specify them differently with `is_writable` or `is_signer`. 

Reviewer: Fixed in latest pushes

### 2. `transfer_hook` Instruction

> In Anchor, when the `Execute` function it looks for and calls the `transfer_hook` instruction. It is the place where we can implement our custom logic for the token transfer.

Reviewer: Fixed in latest pushes

> When the Token Extensions Program CPIs our program, it will invoke this instruction and pass to it all the accounts plus the amount of the transfer that just happened. The first 5 accounts will always be `source`, `mint`, `destination`, `owner`, `extraAccountMetaList`, and the rest are the extra accounts that we added to the `ExtraAccountMetaList` account if there is any.

Reviewer: Fixed in latest pushes

Let's take a look at an example `TransferHook` struct for this instruction:

```rust

// Order of accounts matters for this struct.

// The first 4 accounts are the accounts required for token transfer (source, mint, destination, owner)

// Remaining accounts are the extra accounts required from the ExtraAccountMetaList account

// These accounts are provided via CPI to this program from the Token Extensions Program

#[derive(Accounts)]

pub struct TransferHook<'info> {

  #[account(token::mint = mint, token::authority = owner)]

  pub source_token: InterfaceAccount<'info, TokenAccount>,

  pub mint: InterfaceAccount<'info, Mint>,

  #[account(token::mint = mint)]

  pub destination_token: InterfaceAccount<'info, TokenAccount>,

>   /// CHECK: source token account owner

Reviewer: Fixed in latest pushes

  pub owner: UncheckedAccount<'info>,

>   /// CHECK: ExtraAccountMetaList Account,

Reviewer: Fixed in latest pushes

  #[account(seeds = [b"extra-account-metas", mint.key().as_ref()], bump)]

  pub extra_account_meta_list: UncheckedAccount<'info>,

  // Accounts to add to the extra-account-metas

  pub some_account: UncheckedAccount<'info>,

  #[account(seeds = [b"some-seed"], bump)]

  pub pda_account: UncheckedAccount<'info>,

}

```

As mentioned in the comment, the order here matters; we need the first 5 accounts as shown above, and then the rest of the accounts need to follow the order of the accounts in the `extraAccountMetaList` account.

Other than that, you can write any functionality you want in within the transfer hook. But remember, if the hook fails, the entire transaction fails. 

```rust

  pub fn transfer_hook(ctx: Context<TransferHook>, amount: u64) -> Result<()> {

    // do your logic here

    Ok(())

  }

```

> Note: The transfer hook gets called *after* the transfer happens, so at the point when the transfer hook is getting invoked, the tokens have been transferred from the sender to the receiver.

Reviewer: @jamesrp13

### 3. Fallback

> One last caveat to the on-chain portion of transfer hooks: When dealing with Anchor, we need to specify a `fallback` instruction in the Anchor program to handle the Cross-Program Invocation (CPI) from the Token Extensions Program.

Reviewer: Fixed in latest pushes

This is necessary because Anchor generates instruction discriminators differently from the ones used in the Transfer Hook interface instructions. The instruction discriminator for the `transfer_hook` instruction will not match the one for the Transfer Hook interface.

Next, versions of Anchor should solve this for us, but for now, we can implement this simple workaround:

```rust

// fallback instruction handler as work-around to anchor instruction discriminator check

pub fn fallback<'info>(program_id: &Pubkey, accounts: &'info [AccountInfo<'info>], data: &[u8]) -> Result<()> {

  let instruction = TransferHookInstruction::unpack(data)?;

  // match instruction discriminator to transfer hook interface execute instruction

  // token2022 program CPIs this instruction on token transfer

  match instruction {

      TransferHookInstruction::Execute { amount } => {

      let amount_bytes = amount.to_le_bytes();

      // invoke custom transfer hook instruction on our program

      __private::__global::transfer_hook(program_id, accounts, &amount_bytes)

      }

      _ => {

      return Err(ProgramError::InvalidInstructionData.into());

      }

  }

}

```

> ## Using transfer hooks off-chain

Reviewer: Fixed in latest pushes

> Now that we've looked at the on-chain portion, let's look at how we interact with it off-chain. 

Reviewer: Fixed in latest pushes

Let's assume we have a deployed Solana program that follows the Transfer Hook Interface.

In order to create a mint with a transfer hook and ensure successful transfers, follow these steps:

> 1. Create the mint with the transfer hook extension and point to the on-chain transfer hook program you want to use.

Reviewer: Fixed in latest pushes

    fromPubkey: wallet.publicKey,

    newAccountPubkey: mint.publicKey,

    space: mintLen,

    lamports: lamports,

    programId: TOKEN_2022_PROGRAM_ID,

  }),

  // Initialize the transfer hook extension and point to our program

  createInitializeTransferHookInstruction(

    mint.publicKey,

    wallet.publicKey,

    program.programId, // Transfer Hook Program ID

    TOKEN_2022_PROGRAM_ID,

  ),

  // Initialize mint instruction

  createInitializeMintInstruction(mint.publicKey, decimals, wallet.publicKey, null, TOKEN_2022_PROGRAM_ID),

```

### Initialize `ExtraAccountMetaList` account:

The next step of getting the mint ready for any transactions is initializing the `ExtraAccountMetaList`. Generally, this is done by calling the `initializeExtraAccountMetaList` function on the program containing the transfer hook. Since this is part of the Transfer Hook Interface, this should be standardized. Additionally, if the transfer hook program was made with Anchor, it will most likely have autogenerated IDLs, which are TypeScript interfaces that represent the instructions and accounts of the program. This makes it easy to interact with the program from the client side. 

> If you made your own program in Anchor, you should get the IDLs in the `target/idl` folder. And to make it even easier, if you are inside the anchor project and you are writing tests or client code you can access the methods directly from `anchor.workspace.program_name.method`:

Reviewer: Fixed in latest pushes

## 0. Setup

### 1. Verify Solana/Anchor/Rust Versions

We'll be interacting with the `Token Extensions Program` in this lab and that requires you to have the Solana CLI version ≥ 1.18.1.

To check your version run:

```bash

solana --version

```

If the version printed out after running `solana --version` is less than `1.18.0` then you can update the CLI version manually. Note, at the time of writing this, you cannot simply run the `solana-install update` command. This command will not update the CLI to the correct version for us, so we have to explicitly download version `1.18.0`. You can do so with the following command:

```bash

solana-install init 1.18.1

```

If you run into this error at any point attempting to build the program, that likely means you do not have the correct version of the Solana CLI installed.

```bash

anchor build

error: package `solana-program v1.18.1` cannot be built because it requires rustc 1.72.0 or newer, while the currently active rustc version is 1.68.0-dev

> Either upgrade to rustc 1.72.0 or newer, or use

Reviewer: Fixed in latest pushes

> Either upgrade to rustc 1.72.0 or newer, or use

Reviewer: @jamesrp13

cargo update -p solana-program@1.18.0 --precise ver

where `ver` is the latest version of `solana-program` supporting rustc 1.68.0-dev

```

> You will also want the latest version of the anchor CLI installed. You can follow the steps listed here to update via avm https://www.anchor-lang.com/docs/avm

Reviewer: Fixed in latest pushes

or simply run

```bash

avm install latest

avm use latest

```

At the time of writing, the latest version of the Anchor CLI is `0.29.0`

> 

Reviewer: @jamesrp13

Now, we can check our rust version.

```bash

rustc --version

```

At the time of writing, version `1.76.0` was used for the Rust compiler. If you would like to update, you can do so via `rustup`

https://doc.rust-lang.org/book/ch01-01-installation.html

```bash

rustup update

> ```

Reviewer: @jamesrp13

git checkout starter

```

### 3. Update Program ID and Anchor Keypair

Once in the starter branch, run

```bash

anchor keys sync

```

This syncs your program key with the one in the `Anchor.toml` and the declared program id in the `programs/transfer-hook/src/lib.rs` file. 

The last thing you have to do is set your keypair path in `Anchor.toml`:

```toml

[provider]

cluster = "Localnet"

wallet = "~/.config/solana/id.json"

```

### 4. Confirm the program builds

Let's build the starter code to confirm we have everything configured correctly. If it does not build, please revisit the steps above.

```bash

anchor build

```

You can safely ignore the warnings of the build script, these will go away as we add in the necessary code. But at the end, you should see a message like this:

```bash

Finished release [optimized] target(s)

```

Feel free to run the provided tests to make sure the rest of the dev environment is set up correctly. You'll have to install the node dependencies using `npm` or `yarn`. The tests should run, but they'll all fail until we have completed our program.

```bash

yarn install

anchor test

```

> You should see that 4 tests pass. We will be filling these in later.

Reviewer: Fixed in latest pushes

First, we have to build the struct `InitializeExtraAccountMetaList`, then we can write the instruction itself.

**`InitializeExtraAccountMetaList` Struct**

The Instruction requires the following accounts:

1. `extra_account_meta_list` - The PDA that will hold the extra account.

2. `crumb_mint` - The mint account of the crumb token.

4. `mint` - The mint account of the cookie NFT.

5. `mint_authority` - The mint authority account of the crumb token. - This is a PDA seeded by `b"mint-authority"`

6. `payer` - The account that will pay for the creation of the `extra_account_meta_list` account.

7. `token_program` - The token program account.

8. `system_program` - The system program account.

The code for the struct will go as follows:

```rust

#[derive(Accounts)]

pub struct InitializeExtraAccountMetaList<'info> {

  #[account(mut)]

  payer: Signer<'info>,

  /// CHECK: ExtraAccountMetaList Account, must use these seeds

  #[account(

        mut,

        seeds = [b"extra-account-metas", mint.key().as_ref()], 

        bump

    )]

>   pub extra_account_meta_list: AccountInfo<'info>,

Reviewer: @jamesrp13

  pub mint: InterfaceAccount<'info, Mint>,

  pub token_program: Interface<'info, TokenInterface>,

  pub system_program: Program<'info, System>,

  #[account(mint::authority = mint_authority)]

  pub crumb_mint: InterfaceAccount<'info, Mint>,

  /// CHECK: mint authority Account for crumb mint

  #[account(seeds = [b"mint-authority"], bump)]

  pub mint_authority: UncheckedAccount<'info>,

}

```

Note that we are not specifying the `crumb_mint_ata` or the `associated_token_program`. This is because the `crumb_mint_ata` is variable and will be driven by the other accounts in the `extra_account_meta_list`, and `associated_token_program` will be hardcoded.

> Also, notice we are asking anchor to drive the `mint_authority` account from the seed `b"mint-authority"`. The resulting PDA allows the program itself to sign for the mint.

Reviewer: Fixed in latest pushes

  create_account(

    CpiContext::new(ctx.accounts.system_program.to_account_info(), CreateAccount {

      from: ctx.accounts.payer.to_account_info(),

      to: ctx.accounts.extra_account_meta_list.to_account_info(),

    }).with_signer(signer_seeds),

    lamports,

    account_size,

    ctx.program_id

  )?;

  // Initialize the account data to store the list of ExtraAccountMetas

  ExtraAccountMetaList::init::<ExecuteInstruction>(

    &mut ctx.accounts.extra_account_meta_list.try_borrow_mut_data()?,

    &account_metas

  )?;

  Ok(())

}

```

Pay careful attention to the indexes for each account. Most notably, see that `index 9` is the index for the `crumb_mint_ata` account. It constructs the ATA using `ExtraAccountMeta::new_external_pda_with_seeds` and pass in the seeds from other accounts by their index. Specifically, the ATA belongs to whatever `owner` calls the transfer. So when a cookie is sent, the crumb will be minted to the sender.

### 2. Transfer Hook instruction

In this step, we'll implement the `transfer_hook` instruction. This instruction will be called by the Token Extensions Program when a token transfer occurs. 

> The `transfer_hook` instruction will mint one crumb token each time cookie transfer occurs.
> 

Reviewer 3: ```suggestion
The `transfer_hook` instruction will mint one crumb token each time a cookie transfer occurs.

```

Reviewer: Fixed in latest pushes

Again we'll have a struct `TransferHook` that will hold the accounts required for the instruction.

**`TransferHook` Struct**

In our program the `TransferHook` struct will have 10 accounts:

1. `source_token` - The source token account from which the NFT is transferred.

2. `mint` - The mint account of the Cookie NFT.

3. `destination_token` - The destination token account to which the NFT is transferred.

4. `owner` - The owner of the source token account.

5. `extra_account_meta_list` - The ExtraAccountMetaList account that stores the additional accounts required by the transfer_hook instruction

6. `token_program` - The token program account.

7. `associated_token_program` - The associated token program account.

8. `crumb_mint` - The mint account of the token to be minted by the transfer_hook instruction.

9. `mint_authority` - The mint authority account of the token to be minted by the transfer_hook instruction.

10. `crumb_mint_ata` - The `owner`'s ATA of the crumb mint

> > Very Important Note: The order of accounts in this struct matters. This is the order in which the Token Extensions Program provides these accounts when it CPIs to this Transfer Hook program.

Reviewer 3: ```suggestion
> Very Important Note: The order of accounts in this struct matters. This is the order in which the Token Extensions Program provides these accounts when it invokes this Transfer Hook program.
```

Reviewer: Fixed in latest pushes

        token::authority = owner,

    )]

  pub crumb_mint_ata: InterfaceAccount<'info, TokenAccount>,

}

```

**`transfer_hook` Instruction**

This instruction is fairly simple, it will only make one CPI to the Token Program to mint a new crumb token for each transfer, all that we need to do is to pass the right accounts to the `mint_to` CPI.

Since the mint_authority is a PDA of the transfer hook program itself, the program can sign for it. Therefore we'll use `new_with_signer` and pass mint_authority seeds as the signer seeds.

```rust

pub fn transfer_hook(ctx: Context<TransferHook>, _amount: u64) -> Result<()> {

  let signer_seeds: &[&[&[u8]]] = &[&[b"mint-authority", &[ctx.bumps.mint_authority]]];

  // mint a crumb token for each transaction

  token::mint_to(

    CpiContext::new_with_signer(

      ctx.accounts.token_program.to_account_info(),

      token::MintTo {

        mint: ctx.accounts.crumb_mint.to_account_info(),

        to: ctx.accounts.crumb_mint_ata.to_account_info(),

        authority: ctx.accounts.mint_authority.to_account_info(),

      },

      signer_seeds

    ),

    1

  ).unwrap();

  Ok(())

}

```

> You may have noticed that we are using `token::mint_to` instead of `token_2022::mint_to`, additionally in the `extra_account_meta_list` we're saving the Token Program, not the Token Extensions Program. This is because the crumb SFT *has* to be a Token Program mint and not a Token Extensions Program mint. The answer as to why, is interesting. When first writing this, we wanted to make both the Cookie and Crumb tokens to be Token Extensions Program mints. However, when we did this, we would get a very interesting error: `No Reentrancy`. This happens because the transfer hook is called as a CPI from within the Token Extensions Program, and Solana does not allow [recursive CPIs into the same program](https://defisec.info/solana_top_vulnerabilities). 

Reviewer 3: ```suggestion
You may have noticed that we are using `token::mint_to` instead of `token_2022::mint_to`, additionally in the `extra_account_meta_list` we're saving the Token Program, not the Token Extensions Program. This is because the crumb SFT *has* to be a Token Program mint, not a Token Extensions Program mint. The reason why is interesting: when first writing this, we wanted to make both the Cookie and Crumb tokens to be Token Extensions Program mints. However, when we did this, we would get a very interesting error: `No Reentrancy`. This happens because the transfer hook is called as a CPI from within the Token Extensions Program, and Solana does not allow [recursive CPIs into the same program](https://defisec.info/solana_top_vulnerabilities). 
```

Reviewer: Fixed in latest pushes

To illustrate:

```text

Token Extensions Program -CPI-> Transfer Hook Program -❌CPI❌-> Token Extensions Program

Token Extensions Program -CPI-> Transfer Hook Program -✅CPI✅-> Token Program

```

So, that's why we're making the crumb SFT a Token Program mint.

### 3. Fallback instruction

The last instruction we have to fill out is the `fallback`, this is necessary because Anchor generates instruction discriminators differently from the ones used in Transfer Hook interface instructions. The instruction discriminator for the `transfer_hook` instruction will not match the one for the Transfer Hook interface.

> Next versions of anchor should solve this for us, but for now, we can implement this simple workaround:

Reviewer 3: ```suggestion
Newer versions of Anchor should solve this for us, but for now, we can implement this simple workaround:
```

Reviewer: Fixed in latest pushes

  let instruction = TransferHookInstruction::unpack(data)?;

  // match instruction discriminator to transfer hook interface execute instruction

  // token2022 program CPIs this instruction on token transfer

  match instruction {

      TransferHookInstruction::Execute { amount } => {

      let amount_bytes = amount.to_le_bytes();

      // invoke custom transfer hook instruction on our program

      __private::__global::transfer_hook(program_id, accounts, &amount_bytes)

      }

      _ => {

      return Err(ProgramError::InvalidInstructionData.into());

      }

  }

}

```

### 4. Build the program

Let's make sure our program builds and that tests are runnable before we continue actually writing tests for it.

```bash

anchor test

```

This command will build, deploy and run tests within the `tests/` directory. 

If you're seeing any errors try to go through the steps again and make sure you didn't miss anything.

## 2. Write the tests

Now we'll write some TS scripts to test our code. All of our tests will live inside `tests/anchor.ts`.

The outline of what will we do here is:

> 1. Understand the environment
> 
> 2. Test run all test

Reviewer 3: ```suggestion
1. Understand the environment

2. Run the (empty) tests
```

Reviewer: Fixed in latest pushes

```

Then you can test with:

```bash

anchor test --skip-local-validator

```

### 3. Write the "Create Cookie NFT with Transfer Hook and Metadata" test

Our first test will create our Cookie NFT, which will have metadata and our transfer hook attached.

To accomplish all of this we will create several instructions:

- `SystemProgram.createAccount`: Saves space for the mint on the blockchain

- `createInitializeMetadataPointerInstruction`: Points to the mint itself since the metadata will be stored within the mint

- `createInitializeTransferHookInstruction`: Configures the transfer function to call our transfer hook program

- `createInitializeMintInstruction`: Initializes the mint account

- `createInitializeInstruction`: Adds the metadata to the mint

- `createAssociatedTokenAccountInstruction`: Creates the ATA for the mint to be minted to - owned by the payer

- `createMintToInstruction`: Mints one NFT to the ATA

- `createSetAuthorityInstruction`: Revokes the mint authority, making a true non-fungible token.

Send all of these instructions in a transaction to the blockchain up and you have Cookie NFT:

```ts

> it('Create Cookie NFT with Transfer Hook and Metadata', async () => {

Reviewer 3: ```suggestion
it('Creates a Cookie NFT with Transfer Hook and Metadata', async () => {
```

Reviewer 3: See [`CONTRIBUTING`.](https://github.com/Unboxed-Software/solana-course/blob/main/CONTRIBUTING.md#jsts)

>   // NFT Should have 0 decimals

Reviewer 3: ```suggestion
  // NFTs have 0 decimals
```

### 4. Write the "Create Crumb Mint" test

Now that we have our cookie NFT, we need our crumb SFTs. Creating the crumbs that will be minted on each transfer of our cookie will be our second test.

Remember our crumbs are a Token Program mint, and to attach metadata we need to use Metaplex.

First, we need to grab some Metaplex accounts and format our metadata.

To format our metadata, we need to satisfy Metaplex's `DataV2` struct - for this, we only need to append some additional fields to our `crumbMetadata`.

The Metaplex accounts we will need are:

- `TOKEN_METADATA_PROGRAM_ID`: The Metaplex program

- `metadataPDA`: The metadata account PDA derived from our `crumbMint`

Lastly, to create our crumb, we need the following instructions:

- `SystemProgram.createAccount`: Saves space for our mint

- `createInitializeMintInstruction`: Initializes our mint

- `createCreateMetadataAccountV3Instruction`: Creates the metadata account

- `createSetAuthorityInstruction`: This sets the mint authority to the `crumbMintAuthority`, which is the PDA our transfer hook program owns

Putting it all together we get the following:

```ts

  it('Create Crumb Mint', async () => {

>     // SFT Should have 0 decimals

Reviewer 3: I thought SFTs have more than 0 decimals?

Author: Not necessarily, items in games don't have decimals - but are SFTs. Think of it like a NFT with more than one print

Lastly, after the transfers, we'll grab the crumb mint and assert the total supply is two, and that both the `sourceCrumbAccount` and the `destinationCrumbAccount` have some crumbs.

Putting this all together we get our final test:

```ts

it('Transfer and Transfer Back', async () => {

  const amount = BigInt(1);

  const decimals = 0;

  // Create all of the needed ATAs

  const destinationCookieAccount = (

    await getOrCreateAssociatedTokenAccount(

      connection,

      payerWallet.payer,

      cookieMint.publicKey,

      recipient.publicKey,

      false,

      undefined,

      { commitment: 'confirmed' },

      TOKEN_2022_PROGRAM_ID,

    )

  ).address;

  const sourceCrumbAccount = (

    await getOrCreateAssociatedTokenAccount(

      connection,

      payerWallet.payer,

      crumbMint.publicKey,

      payerWallet.publicKey,

      false,

>       undefined,

Reviewer 3: Use `null` here per previous comments.

Author: Red squiggles

  );

  const transaction = new Transaction().add(transferInstruction, transferBackInstruction);

  const txSig = await sendAndConfirmTransaction(connection, transaction, [payerWallet.payer, recipient], {

    skipPreflight: true,

  });

  console.log(getExplorerLink('transaction', txSig, 'localnet'));

  const mintInfo = await getMint(connection, crumbMint.publicKey, 'processed', TOKEN_PROGRAM_ID);

  const sourceCrumbAccountInfo = await getAccount(connection, sourceCrumbAccount, 'processed', TOKEN_PROGRAM_ID);

  const destinationCrumbAccountInfo = await getAccount(

    connection,

    destinationCrumbAccount,

    'processed',

    TOKEN_PROGRAM_ID,

  );

  expect(Number(mintInfo.supply)).to.equal(2);

  expect(Number(sourceCrumbAccountInfo.amount)).to.equal(1);

  expect(Number(destinationCrumbAccountInfo.amount)).to.equal(1);

  console.log('\nCrumb Count:', Number(mintInfo.supply));

  console.log('Source Crumb Amount:', Number(sourceCrumbAccountInfo.amount));

  console.log('Destination Crumb Amount\n', Number(destinationCrumbAccountInfo.amount));

});

```

Go ahead and run all of the tests:

```bash

anchor test

```

They should all be passing!

> If you want to take a look at any of the explorer links do the following:

Reviewer 3: ```suggestion
If you want to take a look at any of the Explorer links do the following:
```

Author: Done
